src/app/utils/pickupTimeEstimator.js

```diff
@@ -0,0 +1,154 @@
+	let hours = estimatedDeliveryTime.getHours()
+	// round up to the nearest quarter hour, with special accounting for the 
+	// last quarter before the hour, because the hour needs to increment
+	let { hours: roundedHours, minutes: roundedMinutes } = _roundToQuarterHour(hours, minutes)
+
+	if (roundedMinutes < 10) roundedMinutes = `0${roundedMinutes}`
+	const amOrPm = roundedHours > 11 ? 'pm' : 'am'
+	roundedHours = roundedHours % 12 === 0 ? 12 : roundedHours % 12
+	return `${roundedHours}:${roundedMinutes} ${amOrPm}`
+}
+
+// Delivery location should be sc, ma, or my (2 letter codes for each research branch). Request time is a Date object
+export const _expectedAvailableDay = async (deliveryLocation, requestTime, duration) => {
+	const locationHours = await _operatingHours(deliveryLocation)
+	let available
+	let today = new Date(locationHours[0].startTime).getDay()
+	let provisionalDeliveryTime
+	const hours = locationHours.find((day, i) => {
+		// convert everything into ms:
+		const { endTime } = day
+		const endTimeInMs = Date.parse(endTime)
+		provisionalDeliveryTime = Date.parse(requestTime) + duration
+		const requestTimeMs = Date.parse(requestTime)
+		const finalRequestTimeMs = endTimeInMs - REQUEST_CUTOFF_BUFFER
+		//
+
+		// if request was made after request cutoff time, today is not your day
+		if (requestTimeMs > finalRequestTimeMs) return false
+		// if estimated delivery time is before the end of the current day, current
+		// day is the day.
+		if (provisionalDeliveryTime < endTimeInMs) {
+			// determine if that is today, tomorrow, or two days from now.
+			const nextBusinessDay = new Date(provisionalDeliveryTime).getDay()
+			available = _determineNextBusinessDay(today, nextBusinessDay, i)
+			return true
+		}
+	})
+
+	return {
+		available,
+		estimatedDeliveryTime: _calculateDeliveryTime(available, provisionalDeliveryTime, hours.startTime),
+		day: hours.day
+	}
+}
+
+// If it can be delivered today, the provisional estimation still serves.
+// Otherwise, estimated delivery time is the startTime plus the amount
+// of time the branch needs to get their ducks in a row.
+export const _calculateDeliveryTime = (dayAvailable, provisionalDeliveryTime, startTime) => {
+	if (dayAvailable === 'today') return new Date(provisionalDeliveryTime)
+	else {
+		return new Date(Date.parse(startTime) + OPENING_BUFFER)
+	}
+}
+
+export const _determineNextBusinessDay = (today, nextBusinessDay, i) => {
+	if (i === 0) return 'today'
+	// today and nextBusinessDay are one day apart and today is not Saturday
+	if (nextBusinessDay - today === 1) return 'tomorrow'
+	// today is saturday and delivery day is sunday
+	if (today === 6 && nextBusinessDay === 0) return 'tomorrow'
+	// next business day is not tomorrow
+	else return 'two or more days'
+	// TO DO: what happens if the library is closed for a week?
+}
+
+export const _operatingHours = async (deliveryLocation) => {
+	const client = await nyplApiClient()
```

Comment:
We ought to cache operating hours in a local variable for, say, 1hr.

Note that, in practice, `getPickupTimeEstimate` will be called in rapid succession for every item displayed on a page. The first invocation will trigger a call to `_operatingHours`. The second invocation will also trigger a call to `_operatingHours`. The second invocation should not trigger a new HTTP call just because the first hasn't yet resolved. Both invocations should await a single HTTP call and use the same HTTP response once it resolves. You can do this by placing the request Promise in the cache on the first cache-miss (so that the second invocation awaits the first).